src/tiled/scriptdialog.cpp

```diff
@@ -0,0 +1,319 @@
+    return lineEdit;
+}
+QDoubleSpinBox *ScriptDialog::addNumberInput(const QString &labelText)
+{
+    QDoubleSpinBox *doubleSpinBox;
+    checkIfSameType("QDoubleSpinBox");
+    moveToColumn2();
+    if (!labelText.isEmpty()){
+        QLabel * numberLabel = newLabel(labelText);
+        addDialogWidget(numberLabel);
+    }
+    doubleSpinBox = new QDoubleSpinBox(this);
+    addDialogWidget(doubleSpinBox);
+    return doubleSpinBox;
+}
+QSlider *ScriptDialog::addSlider(const QString &labelText)
+{
+    QSlider *horizontalSlider;
+    checkIfSameType("QSlider");
+    moveToColumn2();
+    if(!labelText.isEmpty()){
+        QLabel * sliderLabel= newLabel(labelText);
+        addDialogWidget(sliderLabel);
+    }
+    horizontalSlider = new QSlider(this);
+    horizontalSlider->setOrientation(Qt::Horizontal);
+    horizontalSlider->setContentsMargins(0,0,0,0);
+    addDialogWidget(horizontalSlider);
+    return horizontalSlider;
+}
+
+QCheckBox *ScriptDialog::addCheckBox(const QString &labelText, bool defaultValue)
+{
+    QCheckBox * checkBox;
+    checkIfSameType("QCheckBox");
+    moveToColumn2();
+    checkBox = new QCheckBox(labelText, this);
+    checkBox->setCheckState(defaultValue ? Qt::Checked: Qt::Unchecked);
+    m_gridLayout->addWidget(checkBox, m_rowIndex, m_widgetsInRow);
+    addDialogWidget(checkBox);
+    return checkBox;
+}
+QComboBox *ScriptDialog::addComboBox(const QString &labelText, const QStringList &values)
+{
+    QComboBox * comboBox;
+    checkIfSameType("QComboBox");
+    moveToColumn2();
+    if(!labelText.isEmpty()){
+        QLabel *comboBoxLabel = newLabel(labelText);
+        addDialogWidget(comboBoxLabel);
+    }
+    comboBox = new QComboBox(this);
+    comboBox->addItems(values);
+    m_gridLayout->addWidget(comboBox, m_rowIndex, m_widgetsInRow);
+    addDialogWidget(comboBox);
+    return comboBox;
+}
+
+QPushButton *ScriptDialog::addButton(const QString &labelText)
+{
+    QPushButton *pushButton;
+    checkIfSameType("QPushButton");
+    moveToColumn2();
+    pushButton = new QPushButton(labelText, this);
+    m_gridLayout->addWidget(pushButton, m_rowIndex, m_widgetsInRow);
+    addDialogWidget(pushButton);
+    return pushButton;
+}
+Tiled::FileEdit *ScriptDialog::addFilePicker(const QString &labelText)
+{
+    FileEdit *fileEdit;
+    checkIfSameType("FileEdit");
+    moveToColumn2();
+    if(!labelText.isEmpty()){
+        QLabel *filePickerLabel = newLabel(labelText);
+        addDialogWidget(filePickerLabel);
+    }
+    fileEdit = new FileEdit(this);
+    addDialogWidget(fileEdit);
+    return fileEdit;
+}
+Tiled::ColorButton *ScriptDialog::addColorButton(const QString &labelText)
+{
+    ColorButton *colorButton;
+    checkIfSameType("ColorButton");
+    moveToColumn2();
+    if(!labelText.isEmpty()){
+        QLabel *colorLabel = newLabel(labelText);
+        addDialogWidget(colorLabel);
+    }
+    colorButton = new ColorButton(this);
+    addDialogWidget(colorButton);
+    return colorButton;
```

Comment:
Good idea. Only place it wouldn't work is for things like the color button where I'm editing the size policy to change it from the default after I call addDialogWidget(). But I can do this change soon

Reply:
Added the change to make it return the pointer. Reduces the size of a lot of the methods. thanks

Reply:
Yeah, that's better. I noticed you added casts though, and we actually don't need them, at least not when the function is called from JS. We can just return `QWidget*` also from say `addTextInput` and it will behave exactly the same. In the TypeScript types we should still use the more specific type.